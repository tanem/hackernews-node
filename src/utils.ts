import jwt from 'jsonwebtoken'
import { Context } from './context'

export const APP_SECRET = 'GraphQL-is-aw3some'

export const getUserId = (context: Context) => {
  const Authorization = context.request.get('Authorization')
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '')
    const { userId } = jwt.verify(token, APP_SECRET) as { userId: number }
    return userId
  }

  throw new Error('Not authenticated')
}
