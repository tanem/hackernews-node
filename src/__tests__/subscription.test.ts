import { Link, User } from '@prisma/client'
import bcrypt from 'bcryptjs'
import Chance from 'chance'
import jwt from 'jsonwebtoken'
import path from 'path'
import { SubscriptionClient } from 'subscriptions-transport-ws'
import request from 'supertest'
import { APP_SECRET } from '../utils'

const chance = Chance(path.basename(__filename))

let httpServerUrl: typeof global.__HTTP_SERVER_URL__
let prisma: typeof global.__PRISMA__
let wsServerUrl: typeof global.__WS_SERVER_URL__

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
  wsServerUrl = global.__WS_SERVER_URL__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()
const linkOneDescription = chance.sentence({ words: 5 })
const linkOneUrl = chance.url()
const linkTwoDescription = chance.sentence({ words: 5 })
const linkTwoUrl = chance.url()

let linkId: Link['id']
let token: string
let userId: User['id']

beforeEach(async () => {
  ;({ id: userId } = await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
    },
  }))
  ;({ id: linkId } = await prisma.link.create({
    data: {
      description: linkOneDescription,
      url: linkOneUrl,
      postedBy: {
        connect: { id: userId },
      },
    },
  }))
  token = jwt.sign({ userId }, APP_SECRET)
})

test('new vote', (done) => {
  const client = new SubscriptionClient(wsServerUrl)

  client.onConnected(async () => {
    client
      .request({
        query: `
          subscription {
            newVote {
              id
              link {
                url
                description
              }
              user {
                name
                email
              }
            }
          }
        `,
      })
      .subscribe({
        next: (res) => {
          expect(res.data).toMatchSnapshot()
          done()
        },
      })

    await request(httpServerUrl)
      .post('/')
      .set('Authorization', `Bearer ${token}`)
      .send({
        query: `
          mutation {
            vote(linkId: "${linkId}") {
              id
              link {
                description
                url
              }
              user {
                email
                name
              }
            }
          }
        `,
      })
  })
})

test('new link', (done) => {
  const client = new SubscriptionClient(wsServerUrl)

  client.onConnected(async () => {
    client
      .request({
        query: `
          subscription {
            newLink {
              id
              description
              url
              postedBy {
                id
                name
              }
            }
          }
        `,
      })
      .subscribe({
        next: (res) => {
          expect(res.data).toMatchSnapshot()
          done()
        },
      })

    await request(httpServerUrl)
      .post('/')
      .set('Authorization', `Bearer ${token}`)
      .send({
        query: `
          mutation {
            post(
              url: "${linkTwoUrl}"
              description: "${linkTwoDescription}"
            ) {
              id
              description
              url
              postedBy {
                id
                name
              }
            }
          }
      `,
      })
  })
})
