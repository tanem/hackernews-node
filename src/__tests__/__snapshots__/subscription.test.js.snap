// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new link 1`] = `
Object {
  "newLink": Object {
    "description": "Velaz fo soevu we du.",
    "id": "2",
    "postedBy": Object {
      "id": "1",
      "name": "Mildred",
    },
    "url": "http://lah.cm/casa",
  },
}
`;

exports[`new vote 1`] = `
Object {
  "newVote": Object {
    "id": "1",
    "link": Object {
      "description": "Niwobma sipjatgad pawwopda gu sa.",
      "url": "http://lur.sg/mez",
    },
    "user": Object {
      "email": "behrueri@zu.uk",
      "name": "Mildred",
    },
  },
}
`;
